### ISSUE TITLE
How to undo setText?

### ISSUE THREAD

**Issue by brupelo** [matkuki/qscintilla_docs#3]
@matkuki Hi again! How you doing? Sorry to ping you again... reason being because you're one of the guys I'm aware of that knows quite a lot of QScintilla... so maybe you've got some ideas for this [question](https://stackoverflow.com/questions/54607592/how-to-undo-when-using-qscintilla-settext)?

Thanks in advance ;)

**Comment by matkuki** [comment]
Hey @brupelo ,
No problem, I gave an answer on StackOverflow.
I don't know what you are trying to emulate from SublimeText with the undo/redo functionality, but maybe my answer already solves your problem. If it didn't, give me more information and I'll try my best to give you more ideas.

**Comment by brupelo** [comment]
@matkuki First of all, thank you very much for taking the time to look at the question, really appreciated, I've checked your answer from Stackoverflow and it's definitely a good one, before validating it I'd like to clarify one remaining doubt about scrolling, first of all, please run this new script:

    import textwrap
    import sys
    from pathlib import Path

    from PyQt5.Qsci import QsciScintilla
    from PyQt5.Qt import *  # noqa


    MAX_LINES = 30


    def set_style(sci):
        # Set default font
        sci.font = QFont()
        sci.font.setFamily('Consolas')
        sci.font.setFixedPitch(True)
        sci.font.setPointSize(8)
        sci.font.setBold(True)
        sci.setFont(sci.font)
        sci.setMarginsFont(sci.font)
        sci.setUtf8(True)

        # Set paper
        sci.setPaper(QColor(39, 40, 34))

        # Set margin defaults
        fontmetrics = QFontMetrics(sci.font)
        sci.setMarginsFont(sci.font)
        sci.setMarginWidth(0, fontmetrics.width("000") + 6)
        sci.setMarginLineNumbers(0, True)
        sci.setMarginsForegroundColor(QColor(128, 128, 128))
        sci.setMarginsBackgroundColor(QColor(39, 40, 34))
        sci.setMarginType(1, sci.SymbolMargin)
        sci.setMarginWidth(1, 12)

        # Set indentation defaults
        sci.setIndentationsUseTabs(False)
        sci.setIndentationWidth(4)
        sci.setBackspaceUnindents(True)
        sci.setIndentationGuides(True)
        sci.setFoldMarginColors(QColor(39, 40, 34), QColor(39, 40, 34))

        # Set caret defaults
        sci.setCaretForegroundColor(QColor(247, 247, 241))
        sci.setCaretWidth(2)

        # Set edge defaults
        sci.setEdgeColumn(80)
        sci.setEdgeColor(QColor(221, 221, 221))
        sci.setEdgeMode(sci.EdgeLine)

        # Set folding defaults (http://www.scintilla.org/ScintillaDoc.html#Folding)
        sci.setFolding(QsciScintilla.CircledFoldStyle)

        # Set wrapping
        sci.setWrapMode(sci.WrapNone)

        # Set selection color defaults
        sci.setSelectionBackgroundColor(QColor(61, 61, 52))
        sci.resetSelectionForegroundColor()

        # Set scrollwidth defaults
        sci.SendScintilla(QsciScintilla.SCI_SETSCROLLWIDTHTRACKING, 1)

        # Current line visible with special background color
        sci.setCaretLineBackgroundColor(QColor(255, 255, 224))

        # Set multiselection defaults
        sci.SendScintilla(QsciScintilla.SCI_SETMULTIPLESELECTION, True)
        sci.SendScintilla(QsciScintilla.SCI_SETMULTIPASTE, 1)
        sci.SendScintilla(QsciScintilla.SCI_SETADDITIONALSELECTIONTYPING, True)


    def set_state(sci, num_state, start):
        sci.clear_selections()
        base = "line{:02d} state{}"
        len_base = len(base.format(99, 1)) + 1
        text = "\n".join([base.format(i, num_state) for i in range(MAX_LINES)]).encode("utf-8")
        sci.SendScintilla(sci.SCI_SETTEXT, text)
        # for i in range(start, MAX_LINES, 2):
        #     region = (len_base * i, len_base * (i + 1) - 1)
        #     if i == start:
        #         sci.set_selection(region)
        #     else:
        #         sci.add_selection(region)

        # sci.SendScintilla(sci.SCI_SETMAINSELECTION, -1)


    class Edit():

        def __init__(self, sci):
            self.sci = sci
            self.text = sci.text()
            self.selections = sci.sel()
            self.old_cursor = self.sci.getCursorPosition()

        def insert(self, pt, text):
            offset = len(text)
            # self.sci.insertAt(text, *self.sci.lineIndexFromPosition(pt))
            self.text = self.text[0:pt] + text + self.text[pt:]

            sels = []
            for i, s in enumerate(reversed(self.selections)):
                if s.begin() >= pt:
                    sels.append(Region(s.a + offset, s.b + offset))
                elif s.end() <= pt:
                    sels.append(Region(s.a, s.b))
                else:
                    sels.append(Region(s.a, s.b + offset))

            self.selections = sels

        def flush(self):
            sci = self.sci
            sci.SendScintilla(sci.SCI_SETTEXT, self.text.encode("utf-8"))
            sci.clear_selections()
            for i, s in enumerate(self.selections):
                sci.add_selection(i, s)
            sci.SendScintilla(sci.SCI_SETMAINSELECTION, -1)
            self.sci.setCursorPosition(*self.old_cursor)
            self.sci.setFirstVisibleLine(self.old_cursor[0])


    class Region(object):

        def __init__(self, a, b=None):
            if b is None:
                b = a
            self.a = a
            self.b = b

        def __str__(self):
            return "(" + str(self.a) + ", " + str(self.b) + ")"

        def __repr__(self):
            return "(" + str(self.a) + ", " + str(self.b) + ")"

        def __len__(self):
            return self.size()

        def begin(self):
            if self.a < self.b:
                return self.a
            else:
                return self.b

        def end(self):
            if self.a < self.b:
                return self.b
            else:
                return self.a

        def size(self):
            return abs(self.a - self.b)


    class Editor(QsciScintilla):

        def __init__(self, *args, **kwargs):
            super().__init__(*args, **kwargs)
            set_style(self)

        def clear_selections(self):
            sci = self
            sci.SendScintilla(sci.SCI_CLEARSELECTIONS)

        def set_selection(self, r):
            sci = self
            sci.SendScintilla(sci.SCI_SETSELECTION, r[1], r[0])

         def add_selection(self, i, r):
            if i == 0:
                sci.SendScintilla(sci.SCI_SETSELECTION, r.b, r.a)
            else:
                sci.SendScintilla(sci.SCI_ADDSELECTION, r.b, r.a)

        def sel(self):
            sci = self
            regions = []

            for i in range(sci.SendScintilla(sci.SCI_GETSELECTIONS)):
                regions.append(Region(
                    sci.SendScintilla(sci.SCI_GETSELECTIONNSTART, i),
                    sci.SendScintilla(sci.SCI_GETSELECTIONNEND, i)
                ))

            return sorted(regions)


    if __name__ == '__main__':
        app = QApplication(sys.argv)

        sci = Editor()
        set_state(sci, 1, 0)
        # set_state(sci, 2, 1)

        ed = Edit(sci)
        ed.insert(0, "ABCDE")
        ed.flush()

        sci.move(1000, 100)
        sci.resize(800, 300)
        sci.show()
        app.exec_()

And press `Ctrl+z`, you should get this:

![showcase](https://dl.dropboxusercontent.com/s/1c8sxvw6t2di7d6/2019-02-10_11-40-58.gif)

And now let's take a look at how Sublime behaves... to do that we can use a dummy Sublime command and bind it to some shortcut:

foo.py:

    class FooCommand(sublime_plugin.TextCommand):

        def run(self, edit, block=False):
            # self.view.replace(edit, sublime.Region(0,10), "abcde")
            self.view.insert(edit, 0, "ABCDE")

Default(Windows).sublime-keymap:

    ...
    { "keys": ["ctrl+8"], "command": "foo"},
    ...

If we run it with the same text used on the above snippet we'll see:

![showcase](https://dl.dropboxusercontent.com/s/a7qpoaa1yw2dsl9/2019-02-10_11-35-17.gif)

**QUESTION:** As you can see, the behaviour with Sublime is flawless while with my snippet the scroll is totally broken,
do you know how we can fix it?

> I don't know what you are trying to emulate from SublimeText with the undo/redo functionality, but maybe my answer already solves your problem. If it didn't, give me more information and I'll try my best to give you more ideas.

Thing is, I'm in the middle of porting the insert/replace/erase Sublime functionality to QScintilla, that way I'll be able to use many Sublime commands directly on QScintilla but first I'm trying to sort out the issue with the scrolling as well as figure out how `replace` [works](https://forum.sublimetext.com/t/whats-the-view-replace-algorithm/41781) in SublimeText

PS: Btw, you can see in the insert method of the Edit class I'd tried insertAt, which works decently well (when undoing)... but i'd like to avoid scintilla commands when modifying the internal buffer/text... i think that will be faster when doing many edit operations (insert/erase/replace)

**Comment by matkuki** [comment]
@brupelo,
`Sublime` does not behave the same, as you are using a keyboard shortcut to insert text using `Sublime`'s equivalent to `QScintilla`'s `insertAt`, while you are using your manual code in the `QScintilla` example. It works the same in `QScintilla` (`Editor` in your case) if you add a shortcut function using `insertAt`:
```python
def keyReleaseEvent(self, event):
        super().keyReleaseEvent(event)
        key = event.key()
        key_modifiers = QApplication.keyboardModifiers()
        if key_modifiers == Qt.ControlModifier:
            if key == Qt.Key_I:
                self.insertAt("ABCDE", 0, 0)
```

What do you mean you do not want to use the internal buffer/text? That is unavoidable. Even when you use `sci.SendScintilla(sci.SCI_SETTEXT, self.text.encode("utf-8"))` in your `flush` function, you are modifying the internal buffer. And in this case `sci.SCI_SETTEXT` **ALWAYS** scrolls to the bottom.

But if you want to stick with the `flush` method with `sci.SCI_SETTEXT`, you need to add a `cache` to store `cursor` and `firstVisibleLine` on every `insert`/`flush`/... , then override `Editor`'s `undo` and `redo` functions and add code to correctly position the cursor and first visible line on every undo or redo.

**Comment by brupelo** [comment]
> But if you want to stick with the `flush` method with `sci.SCI_SETTEXT`, you need to add a `cache` to store `cursor` and `firstVisibleLine` on every `insert`/`flush`/... , then override `Editor`'s `undo` and `redo` functions and add code to correctly position the cursor and first visible line on every undo or redo.

This! 

As you can see in the above snippet I'd already tried using `setCursorPosition` and `setFirstVisibleLine` with no success?

    def flush(self):
        sci = self.sci
        sci.SendScintilla(sci.SCI_SETTEXT, self.text.encode("utf-8"))
        sci.clear_selections()
        for i, s in enumerate(self.selections):
            sci.add_selection(i, s)
        sci.SendScintilla(sci.SCI_SETMAINSELECTION, -1)
        self.sci.setCursorPosition(*self.old_cursor)
        self.sci.setFirstVisibleLine(self.old_cursor[0])

Also.. what do you mean by overriding `undo` & `redo`? The idea is this, when you're creating a new command that mimicks Sublime's at certain point you'll have a bunch of "atomic" operations:

    # Checkpoint1
    ed = Edit(sci) 
    ed.insert(...)
    ed.replace(...)
    ed.erase(...)
    ...
    ed.flush()  
    # Checkpoint2

So basically after flushing (Checkpoint2/state), if you press ctrl+z you'd be able to get the state of Checkpoint1 (with proper scrolling).

Not sure if `beginUndoAction` and `endUndoAction` could help us here though... In any case, I haven't made it work yet... Have you?

Thanks in advance :)

Ps. The goal? Once I've got these atomic operations working in place I'll come back to some old issue that was commenting/uncommenting blocks... it should be possible to use directly Sublime's code.

**Comment by matkuki** [comment]
> As you can see in the above snippet I'd already tried using `setCursorPosition` and `setFirstVisibleLine` with no success?

Calls on `setCursorPosition` and `setFirstVisibleLine` are not stored in the internal buffer. Which means that when you `flush`, the cursor and first line are set to the correct position (as they are in your example), but then when you `undo` with `Ctrl+Z` the cursor and first line move to the position of what `sci.SendScintilla(sci.SCI_SETTEXT, self.text.encode("utf-8"))` sets it to, which is the end of the text. To make it work how you want, you would then need manual `undo` and `redo` functions. First you would need to create a manual `undo` and `redo` function that undo plus position the cursor and first line correctly, then remove the `Ctrl+Z` and `Ctrl+Y` keyboard shotcuts by using `sci.SCI_CLEARCMDKEY`, and then add manual calls to your new functions for `Ctrl+Z` and `Ctrl+Y` like I did for the `Ctrl+I` example above in the  overridden `keyReleaseEvent` function.

> Not sure if `beginUndoAction` and `endUndoAction` could help us here though... In any case, I haven't made it work yet... Have you?

These functions are used when you want to group multiple actions into one `undo/redo` operation. For example when you call this code:
```
self.beginUndoAction()
self.insertAt("ABCDE", 0, 0)
self.insertAt("FGHIJ", 0, 0)
self.insertAt("KLMNO", 0, 0)
self.endUndoAction()
```
When pressing `Ctrl+Z` **once**, the editor's text is set back to the state before the first call to `insertAt`. This is so you don't need to press `Ctrl+Z` three times in order to get back to the previous state.

**Comment by brupelo** [comment]
Thanks... wow, sounds complicated hehe :)

Considering your explanation I think I'll end up using the QScintilla working out of the box commands then :/... seems that way I'll have it working more easily... if eventually the performance is not good I'll be worry about it later.

That said, do you know how to implement [replace](https://forum.sublimetext.com/t/whats-the-view-replace-algorithm/41781)?

Ps. validated your SO answer already, you've made more than enough, as usual, tyvm!

**Comment by matkuki** [comment]
>That said, do you know how to implement replace?

Looking at you tests animation, it seems that when doing **one** selection it preserves the selection of the new region back-to-front. You could try replacing the region with a string bigger that the initial text to get more information.

As for **multiple** selections, I have no idea what's going on. It seems it just tries to re-position to the subtraction of the sizes `pos = pos - (len("0123456789") - len("abcde"))`.

Good luck

**Comment by brupelo** [comment]
@matkuki Sorry to come back to this issue but I've realized being able to restore properly Scintilla "States" is actually important if you want to have a proper editor supporting multiselection, let me start by putting you an example, consider this Sublime command:

    class ZebraFooCommand(sublime_plugin.TextCommand):

        def run(self, edit, block=False):
            self.view.insert(edit, 3, "hello")
            self.view.insert(edit, 10, "world")
            self.view.erase(edit, sublime.Region(0,2))
            self.view.replace(edit, sublime.Region(15,20), "foobar\n"*200)

If we run this sublime command few times, the state of the editor {selections, first_visible_line, text buffer} will be saved before and after its execution. Let me show you:

![showcase](https://dl.dropboxusercontent.com/s/xt6fs8024lkxsdv/2019-02-13_12-09-51.gif)

Problem is default Scintilla undo/redo won't work that way... if we check the [Scintilla docs](https://www.scintilla.org/ScintillaDoc.html#UndoAndRedo):

> Scintilla has multiple level undo and redo. It will continue to collect undoable actions until memory runs out. Scintilla saves actions that change the document. Scintilla does not save caret and selection movements, view scrolling and the like. Sequences of typing or deleting are compressed into single transactions to make it easier to undo and redo at a sensible level of detail. Sequences of actions can be combined into transactions that are undone as a unit. These sequences occur between SCI_BEGINUNDOACTION and SCI_ENDUNDOACTION messages. These transactions can be nested and only the top-level sequences are undone as units.

Just for the sake of being more concise when talking about this topic, let's add a little bit of notation about what "States" and "Commands" are:

    S=State; {set of selections, first_visible_line, buffer_text}
    S{i} = State indexed by i, ie: S0 (state0), S1(state1), ..., Sn(stateN)
    C=Command; {collection of insert/erase/replace operations}
    C{i} = Command indexed by i, ie: C0 (command0), C1(command1), ..., Cn(commandN)

Example:

    S0->C0->S1->C1->S2->...Cn->Sn

In the above example you can see how you commands allow you to transition between states, so basically when you undo/redo in Sublime you're just moving across contiguous states.

Anyway, that said, the goal here is being able to override undo/redo somehow to mimick Sublime's behaviour and I think you've got the idea in your idea how that could be implemented, let's consider this mcve:

    import textwrap
    import sys
    from pathlib import Path

    from PyQt5.Qsci import QsciScintilla
    from PyQt5.Qt import *  # noqa


    def set_style(sci):
        # Set default font
        sci.font = QFont()
        sci.font.setFamily('Consolas')
        sci.font.setFixedPitch(True)
        sci.font.setPointSize(8)
        sci.font.setBold(True)
        sci.setFont(sci.font)
        sci.setMarginsFont(sci.font)
        sci.setUtf8(True)

        # Set paper
        sci.setPaper(QColor(39, 40, 34))

        # Set margin defaults
        fontmetrics = QFontMetrics(sci.font)
        sci.setMarginsFont(sci.font)
        sci.setMarginWidth(0, fontmetrics.width("000") + 6)
        sci.setMarginLineNumbers(0, True)
        sci.setMarginsForegroundColor(QColor(128, 128, 128))
        sci.setMarginsBackgroundColor(QColor(39, 40, 34))
        sci.setMarginType(1, sci.SymbolMargin)
        sci.setMarginWidth(1, 12)

        # Set indentation defaults
        sci.setIndentationsUseTabs(False)
        sci.setIndentationWidth(4)
        sci.setBackspaceUnindents(True)
        sci.setIndentationGuides(True)
        sci.setFoldMarginColors(QColor(39, 40, 34), QColor(39, 40, 34))

        # Set caret defaults
        sci.setCaretForegroundColor(QColor(247, 247, 241))
        sci.setCaretWidth(2)

        # Set edge defaults
        sci.setEdgeColumn(80)
        sci.setEdgeColor(QColor(221, 221, 221))
        sci.setEdgeMode(sci.EdgeLine)

        # Set folding defaults (http://www.scintilla.org/ScintillaDoc.html#Folding)
        sci.setFolding(QsciScintilla.CircledFoldStyle)

        # Set wrapping
        sci.setWrapMode(sci.WrapNone)

        # Set selection color defaults
        sci.setSelectionBackgroundColor(QColor(61, 61, 52))
        sci.resetSelectionForegroundColor()

        # Set scrollwidth defaults
        sci.SendScintilla(QsciScintilla.SCI_SETSCROLLWIDTHTRACKING, 1)

        # Current line visible with special background color
        sci.setCaretLineBackgroundColor(QColor(255, 255, 224))

        # Set multiselection defaults
        sci.SendScintilla(QsciScintilla.SCI_SETMULTIPLESELECTION, True)
        sci.SendScintilla(QsciScintilla.SCI_SETMULTIPASTE, 1)
        sci.SendScintilla(QsciScintilla.SCI_SETADDITIONALSELECTIONTYPING, True)


    def set_state1(sci):
        sci.clear_selections()
        base = "line{} state1"
        view.setText("\n".join([base.format(i) for i in range(10)]))
        for i in range(0, 10, 2):
            region = (len(base) * i, len(base) * (i + 1) - 1)
            if i == 0:
                view.set_selection(region)
            else:
                view.add_selection(region)


    def set_state2(sci):
        base = "line{} state2"
        view.setText("\n".join([base.format(i) for i in range(10)]))
        for i in range(1, 10, 2):
            region = (len(base) * i, len(base) * (i + 1) - 1)
            if i == 1:
                view.set_selection(region)
            else:
                view.add_selection(region)


    class Editor(QsciScintilla):

        def __init__(self, *args, **kwargs):
            super().__init__(*args, **kwargs)
            set_style(self)

        def clear_selections(self):
            sci = self
            sci.SendScintilla(sci.SCI_CLEARSELECTIONS)

        def set_selection(self, r):
            sci = self
            sci.SendScintilla(sci.SCI_SETSELECTION, r[1], r[0])

        def add_selection(self, r):
            sci = self
            sci.SendScintilla(sci.SCI_ADDSELECTION, r[1], r[0])

        def sel(self):
            sci = self
            regions = []

            for i in range(sci.SendScintilla(sci.SCI_GETSELECTIONS)):
                regions.append(
                    sci.SendScintilla(sci.SCI_GETSELECTIONNSTART, i),
                    sci.SendScintilla(sci.SCI_GETSELECTIONNEND, i)
                )

            return sorted(regions)


    if __name__ == '__main__':
        app = QApplication(sys.argv)

        view = Editor()
        set_state1(view)
        set_state2(view)
        view.move(1000, 100)
        view.resize(800, 300)
        view.show()
        app.exec_()

How would you modify this example so undo/redo will work like Sublime's?

PS: When looking at the gif example, you can see i run the command and I undo... sometimes by just looking at the main view the behaviour may be not clear... but if you look at the minimap you'll perfectly realize the sequence I've followed on that demo

**Comment by brupelo** [comment]
Btw, here's another simpler example to give more background about my previous question so you'll realize a little bit more how ST works, ie:

![showcase](https://dl.dropboxusercontent.com/s/l6vgvfmhuockxew/2019-02-13_12-42-00.gif)
